libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt

```diff
@@ -159,8 +166,31 @@ abstract class KaptTask : ConventionTask(), TaskWithLocalState {
     @Internal
     protected fun getCompiledSources() = listOfNotNull(kotlinCompileTask.destinationDir, kotlinCompileTask.javaOutputDir)
 
-    protected fun getChangedFiles(inputs: IncrementalTaskInputs): List<File> {
-        return if (!isIncremental || !inputs.isIncremental || !getCompiledSources().all { it.exists() }) {
+    protected fun getIncrementalChanges(inputs: IncrementalTaskInputs): KaptIncrementalChanges {
+        val changedFiles = getChangedFiles(inputs)
+
+        return if (isIncremental) {
+            val classpathChanges = if (changedFiles.isEmpty()) {
+                classpath.files
+            } else {
+                classpath.files.let { cp ->
```

Comment:
In case of a removed classpath entry, `ClasspathSnapshot.diff` will return early because previous and current classpath are not compatible. Right now, 2 classpath are compatible if they contain the same entries, in the same order. Handling addition/removals/swaps requires additional work.